sdv/sampling/hierarchical_sampler.py

```diff
@@ -0,0 +1,210 @@
+                    parent_row=row,
+                    sampled_data=sampled_data
+                )
+
+    def _finalize(self, sampled_data):
+        """Remove extra columns from sampled tables and apply finishing touches.
+
+        This method reverts the previous transformations to go back
+        to values in the original space.
+
+        Args:
+            sampled_data (dict):
+                Dictionary mapping table names to sampled tables (pd.DataFrame)
+
+        Returns:
+            Dictionary mapping table names to their formatted sampled tables.
+        """
+        final_data = {}
+        for table_name, table_rows in sampled_data.items():
+            synthesizer = self._table_synthesizers.get(table_name)
+            dtypes = synthesizer._data_processor._dtypes
+            for name, dtype in dtypes.items():
+                table_rows[name] = table_rows[name].dropna().astype(dtype)
+
+            final_data[table_name] = table_rows[list(dtypes.keys())]
+
+        return final_data
+
+    def _sample(self, scale=1.0):
+        """Sample the entire dataset.
+
+        Returns a dictionary with all the tables of the dataset. The amount of rows sampled will
+        depend from table to table. This is because the children tables are created modelling the
+        relation that they have with their parent tables, so its behavior may change from one
+        table to another.
+
+        Args:
+            scale (float):
+                A float representing how much to scale the data by. If scale is set to ``1.0``,
+                this does not scale the sizes of the tables. If ``scale`` is greater than ``1.0``
+                create more rows than the original data by a factor of ``scale``.
+                If ``scale`` is lower than ``1.0`` create fewer rows by the factor of ``scale``
+                than the original tables. Defaults to ``1.0``.
+
+        Returns:
+            dict:
+                A dictionary containing as keys the names of the tables and as values the
+                sampled data tables as ``pandas.DataFrame``.
+        """
+        sampled_data = {}
+        for table in self.metadata.tables:
+            if not self.metadata._get_parent_map().get(table):
+                num_rows = int(self._table_sizes[table] * scale)
+                synthesizer = self._table_synthesizers[table]
+                self._sample_table(
+                    synthesizer=synthesizer,
+                    table_name=table,
+                    num_rows=num_rows,
+                    sampled_data=sampled_data
+                )
+
+        for relationship in self.metadata.relationships:
+            parent_name = relationship['parent_table_name']
+            child_name = relationship['child_table_name']
+            self._add_foreign_key_columns(
+                sampled_data[child_name],
+                sampled_data[parent_name],
+                child_name,
+                parent_name
+            )
```

Comment:
Should we check if the foreign key column already exists?

Reply:
I think since we're adding all the foreign key columns in  the `_add_foreign_key_columns` function, it makes sense for it to handle that check. We should make sure we're only calling that function once per relationship though, so I fixed that.